File: Blind75/ValidAnagram/ValidAnagram.py
#! 242. Valid Anagram

# Given two strings s and t, return true if t is an anagram of s, and false otherwise.
# An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

# Example 1:
# Input: s = "anagram", t = "nagaram"
# Output: true

# Example 2:
# Input: s = "rat", t = "car"
# Output: false

# Constraints:
# 1 <= s.length, t.length <= 5 * 104
# s and t consist of lowercase English letters.

class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        # Set edge case to check if s & t are same length,
        if len(s) != len(t):
            return False

        # Create empty hashmaps for each string (dictionaries)
        countS, countT = {},{}

        # Loop through the length of s, increment the count
        # of each charecter in the hashmap
        for i in range(len(s)):
            countS[s[i]] = 1 + countS.get(s[i],0)
            countT[t[i]] = 1 + countT.get(t[i],0)
            # .get returns the current count of the charecter, or 0
            # if that charecter is not yet mapped. Add 1 to count.

        # Testing to see if the counts are same across hashmaps
        for c in countS:
            # Check if the count of character c in s is not equal
            # to the count of character c in t. If not, return False.
            if countS[c] != countT.get(c,0):
                return False

        # Return True if it passes all tests.
        return True
    
# Test Cases
solution = Solution()
# Example 1, expected output = true
print(solution.isAnagram("anagram","nagaram"))

# Example 2, expected output = false
print(solution.isAnagram("rat","car"))

# Example 3, expected output = false
print(solution.isAnagram("a","ab"))

# Example 4, expected output = true
print(solution.isAnagram("god","dog"))

File: Blind75/ValidAnagram/ValidAnagram.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */

var isAnagram = function (s, t) {
    console.log(`Checking if String ("${s}") and String ("${t}") are Anagrams.`)

    // Set-up edge case for if strings are not equal length
    if (s.length !== t.length) {
        console.log(`String ("${s}") and String ("${t}") are not equal length; therefore they are not Anagrams.`)
        return false;
    }

    // Set up a hashmap to push chars into from string (s) and delete from with string (t)
    const hashMap = {};

    // Loop through string (s) and push char into hashmap
    for (i = 0; i < s.length; i++) {
        let c = s[i];
        hashMap[c] = (hashMap[c] || 0) + 1;
        console.log(`Adding an increment to the count of ${c}`)
        console.log(hashMap)
    }

    // Now that our hashmap is populated with the (s) string, it's time to revert it to empty with the (t) string
    for (i = 0; i < t.length; i++) {
        let c = t[i];
        // Check if c exists in hashmap, if not => break loop and return false
        if (!hashMap[c]) {
            console.log(`The count for char ${c} is invalid, these are not Anagrams.`);
            return false
        } else {
            // If it does, decrement the count of the char by 1 
            hashMap[c]--;
            console.log(`Decreasing count of char ${c} in Hashmap`)
            console.log(hashMap)
        }
    }
    console.log("These are valid Anagrams")
    return true

};

// Tests:
console.log(isAnagram('anagram', 'nagaram')); // true
console.log("\n")
console.log(isAnagram('rat', 'car')); // false
console.log("\n")
console.log(isAnagram('a', 'ab')); // false